Bring fast_nvcc.py to PyTorch OSS

samestep opened the pull request:
Summary: This PR adds `tools/fast_nvcc/fast_nvcc.py`, a mostly-transparent wrapper over `nvcc` that parallelizes compilation of CUDA files when building for multiple architectures at once.

Test Plan: Build PyTorch as usual (with CUDA); it should still work and hopefully also be faster.

Differential Revision: D25286030

walterddr commented on tools/README.md:
```diff
@@ -29,6 +29,8 @@ Build system pieces:
 * [build_libtorch.py](build_libtorch.py) - Script for building
   libtorch, a standalone C++ library without Python support.  This
   build script is tested in CI.
+* [fast_nvcc](fast_nvcc) - Mostly-transparent wrapper script that parallelizes
```
i believe it follows markdown syntax (e.g. `[link_display_name](link_relative_path)`
```suggestion
* [fast_nvcc.py](fast_nvcc/fast_nvcc.py) - Mostly-transparent wrapper script that parallelizes
```

samestep replied:
I'd rather leave it as-is to match the existing convention in this README (e.g. for `docker/Dockerfile_runtime`). If you want, though, I could change it to a bullet with a subbullet, similar to what's currently done for `amd_build/build_amd.py` and `shared/module_loader.py`. Preference?

walterddr replied:
oh. I see. are you planning to add more to the folder `fast_nvcc`? if yes, then the current bullet `[fast_nvcc]` looks good! and yes if you can let's add a subbullet for `fast_nvcc/fast_nvcc.py`.

samestep replied:
I don't currently have any plans to add more files to that folder, but I needed it to be in its own folder for hg dirsync to work; but yea I'll add a subbullet :)

samestep committed:
Bring fast_nvcc.py to PyTorch OSS (#48934)

Summary:
This PR adds `tools/fast_nvcc/fast_nvcc.py`, a mostly-transparent wrapper over `nvcc` that parallelizes compilation of CUDA files when building for multiple architectures at once.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/48934

Test Plan: Currently this script isn't actually used in PyTorch OSS. Coming soon!

Differential Revision: D25286030

Pulled By: samestep

fbshipit-source-id: c8e76a1ece5d74717e630ee1b79d99651e735015

vors commented:
Neat idea! Did you measure the compilation speed up by any chance?

samestep commented:
> Neat idea! Did you measure the compilation speed up by any chance?

@vors I didn't personally measure it for our OSS builds, but @walterddr did in #49773 :)
